ocaml/xapi/xapi_pool_update.ml

```diff
@@ -43,6 +44,124 @@ type update_info = {
   hosts: API.ref_host list
 }
 
+(** Mount a filesystem somewhere, with optional type *)
+let mount ?ty:(ty = None) ?lo:(lo = true) src dest =
+  let ty = match ty with None -> [] | Some ty -> [ "-t"; ty ] in
+  let lo = if lo then ["-o"; "loop"] else [] in
+  ignore(Forkhelpers.execute_command_get_output "/bin/mount" (ty @ lo @ [src; dest ]))
+
+let timeout = 300. (* 5 minutes: something is seriously wrong if we hit this timeout *)
+exception Umount_timeout
+
+(** Unmount a mountpoint. Retries every 5 secs for a total of 5mins before returning failure *)
+let umount ?(retry=true) dest =
+  let finished = ref false in
+  let start = Unix.gettimeofday () in
+
+  while not(!finished) && (Unix.gettimeofday () -. start < timeout) do
+    try
+      ignore(Forkhelpers.execute_command_get_output "/bin/umount" [dest] );
+      finished := true
+    with e ->
+      if not(retry) then raise e;
+      debug "Caught exception (%s) while unmounting %s: pausing before retrying"
+        (ExnHelper.string_of_exn e) dest;
+      Thread.delay 5.
+  done;
+  if not(!finished) then raise Umount_timeout
+
+let detach ~__context ~self ~host =
+  let uuid = Db.Pool_update.get_uuid ~__context ~self in
+  let mount_point_parent_dir = String.concat "/" [Xapi_globs.host_update_dir; uuid] in
+  let mount_point = String.concat "/" [mount_point_parent_dir; "vdi"] in
+  let vdi = Db.Pool_update.get_vdi ~__context ~self in
+  let sr = Db.VDI.get_SR ~__context ~self:vdi in
+  let shared = Db.SR.get_shared ~__context ~self:sr in
+  let pbds = Db.SR.get_PBDs ~__context ~self:sr in
+  let plugged_pbds = List.filter (fun pbd -> Db.PBD.get_currently_attached ~__context ~self:pbd) pbds in
+  let plugged_hosts = List.setify (List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd) plugged_pbds) in
+  debug "pool_update.detach %s from %s" (Db.Pool_update.get_name_label ~__context ~self) mount_point;
+  if (shared = true) || ((List.length pbds = 1) && (List.exists ( fun h -> h = host) plugged_hosts)) then begin
+    umount mount_point;
```

Comment:
If you call this API call (Pool_update.detach) without having attached first, this call to umount will always block for 5 minutes here, right? If so, we should make sure we catch this condition.